### PR TITLE
Secret input auto complete

### PR DESCRIPTION
**Pull request by Salman2301**
# Description 📣
Improve secret input component auto complete, lint and error validations
- [x] Move to react query
- [x] Use Radix component for portal, scroll and keyboard navigation popup
- [ ] Refactor `onChange`?
- [x] Fix popup position on overview (Maybe move to radix component will fix it)
- [x] Invalidate Input on improper reference
- [x] Escape [dot] character
- [ ] ~Highlight different color for env, folders and secrets?~ (Different color can be too heavy skipping for now)
- [x] invalidate `/` `.` in environment slug 

## Type ✨

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation

# Tests 🛠️

https://github.com/Infisical/infisical/assets/40920317/1dfbb38b-3a98-45a1-a302-7bb4deeb426d

```sh
# Here's some code block to paste some code snippets
```

---

- [x] I have read the [contributing guide](https://infisical.com/docs/contributing/getting-started/overview), agreed and acknowledged the [code of conduct](https://infisical.com/docs/contributing/getting-started/code-of-conduct). 📝

### CODE REVIEW

**Comment by Salman2301** [comment]
This is ready for review

**Comment by Salman2301** [comment]
The UI in the video is kind of outdated after moved to Radix component.
Will update it later

**Comment by Salman2301** [comment]
The UI in the video is kind of outdated after moved to Radix component.
Will update it later

**Comment by Salman2301** [comment]
Using Radix component... looks better in secret overview cause of portal I believe but input get out of focus on this version everytime it's popup
https://github.com/Infisical/infisical/assets/40920317/ea0e6303-92f8-43c4-bcc4-571657b9c2f6

**Comment by maidul98** [comment]
@Salman2301 wow this looks great! @akhilmhdh can you please help @Salman2301 get this into `main`?

File: backend/src/server/routes/v1/project-env-router.ts [hunk 1]
```diff
@@ -22,7 +22,10 @@ export const registerProjectEnvRouter = async (server: FastifyZodProvider) => {
       }),
       body: z.object({
         name: z.string().trim(),
-        slug: z.string().trim()
+        slug: z
```
Comment (akhilmhdh) [inline backend/src/server/routes/v1/project-env-router.ts:25]: Use slugify for slug validation. You can find it used in multiple places

File: frontend/src/components/v2/SecretInput/SecretInput.tsx [hunk 2]
```diff
@@ -1,10 +1,15 @@
 /* eslint-disable react/no-danger */
-import { forwardRef, TextareaHTMLAttributes } from "react";
+import React, { forwardRef, TextareaHTMLAttributes, useRef, useState } from "react";
 import { twMerge } from "tailwind-merge";
 
+import {
```
Comment (akhilmhdh) [inline frontend/src/components/v2/SecretInput/SecretInput.tsx:5]: Let's keep this as a different component. Secret input is used in multiple places where secret reference is not needed. May be name the new component as `InfisicalSecretInput`. We can rename this component later

File: frontend/src/components/v2/SecretInput/SecretInput.tsx [hunk 3]
```diff
@@ -1,10 +1,15 @@
...
+  REGEX_SECRET_REFERENCE_FIND,
```
Comment (akhilmhdh) [inline frontend/src/components/v2/SecretInput/SecretInput.tsx:6]: Avoid distributing the logics into so many files. This logic is never shared anywhere other than mostly thing component so keep the regex here, co-location functions is easier to manage

File: frontend/src/components/v2/SecretInput/SecretInput.tsx [hunk 4]
```diff
@@ -49,47 +63,259 @@ type Props = TextareaHTMLAttributes<HTMLTextAreaElement> & {
...
+        childRef.current.focus();
+        setTimeout(() => {
+          if (!childRef?.current) return;
+          childRef.current.selectionStart = caretPos;
+          childRef.current.selectionEnd = caretPos;
+        }, 200);
+      }
+    }
+
+    function referencePopup(text: string, pos: number) {
```
Comment (akhilmhdh) [inline frontend/src/components/v2/SecretInput/SecretInput.tsx:123]: Kindly switch to arrow function

File: frontend/src/components/v2/SecretInput/SecretInput.tsx [hunk 5]
```diff
@@ -49,47 +63,259 @@ type Props = TextareaHTMLAttributes<HTMLTextAreaElement> & {
...
+          return match;
+        }
+      }
+      return null;
+    }
+
+    function setCaretPos(caretPos: number) {
+      if (childRef?.current) {
+        childRef.current.focus();
+        setTimeout(() => {
```
Comment (akhilmhdh) [inline frontend/src/components/v2/SecretInput/SecretInput.tsx:115]: setTimeout is missing `clearTimeout` in turn causes memory leak

File: frontend/src/views/SecretRotationPage/components/CreateRotationForm/steps/RotationInputForm.tsx [hunk 6]
```diff
@@ -13,11 +13,13 @@ type Props = {
     properties: Record<string, { type: string; desc?: string; default?: string }>;
     required: string[];
   };
+  secretPath: string;
+  environment: string;
 };
 
 const formSchema = z.record(z.string().trim().optional());
 
-export const RotationInputForm = ({ onSubmit, onCancel, inputSchema }: Props) => {
```
Comment (akhilmhdh) [inline frontend/src/views/SecretRotationPage/components/CreateRotationForm/steps/RotationInputForm.tsx:20]: By keeping it in another component you can discard these changes

File: frontend/src/views/Settings/ProjectSettingsPage/components/EnvironmentSection/AddEnvironmentModal.tsx [hunk 7]
```diff
@@ -16,7 +16,7 @@ type Props = {
 
 const schema = yup.object({
   environmentName: yup.string().label("Environment Name").required(),
-  environmentSlug: yup.string().label("Environment Slug").required()
+  environmentSlug: yup.string().label("Environment Slug").matches(/^[^./]*$/g, { message: "Invalid [.] or [/] not allowed"}).required()
```
Comment (akhilmhdh) [inline frontend/src/views/Settings/ProjectSettingsPage/components/EnvironmentSection/AddEnvironmentModal.tsx:19]: Use slugify lib for validation

File: frontend/src/components/v2/SecretInput/SecretInput.tsx [hunk 8]
```diff
@@ -49,47 +63,259 @@ type Props = TextareaHTMLAttributes<HTMLTextAreaElement> & {
...
+          break;
+        default:
+      }
+      replaceReference = replaceReference.replace(/[//]/g, "");
+      newValue = `${start}$\{${replaceReference}}${end}`;
+      setValue(newValue);
+      // TODO: there should be a better way to do
+      onChange?.({ target: { value: newValue } } as any);
+      setShowReferencePopup(type !== "secret");
+      setTimeout(() => {
```
Comment (akhilmhdh) [inline frontend/src/components/v2/SecretInput/SecretInput.tsx:246]: Same comment on previous timeout. Handle clearTimeout

File: frontend/src/components/v2/SecretInput/SecretReferenceSelect.tsx [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+import {
+  faCaretDown,
+  faCaretUp,
+  faChevronRight,
+  faFolder,
+  faKey,
+  faRecycle
+} from "@fortawesome/free-solid-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import * as SelectPrimitive from "@radix-ui/react-select";
```
Comment (akhilmhdh) [inline frontend/src/components/v2/SecretInput/SecretReferenceSelect.tsx:11]: Use `Select` component and not radix-select

File: frontend/src/components/v2/SecretInput/SecretReferenceSelect.tsx [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+import { twMerge } from "tailwind-merge";
+
+import { useWorkspace } from "@app/context";
+import { useGetUserWsKey } from "@app/hooks/api";
+import { useGetFoldersByEnv } from "@app/hooks/api/secretFolders/queries";
+import { useGetProjectSecrets } from "@app/hooks/api/secrets/queries";
+
+type ReferenceType = "environment" | "folder" | "secret";
```
Comment (akhilmhdh) [inline frontend/src/components/v2/SecretInput/SecretReferenceSelect.tsx:19]: Use enum please

**Review by akhilmhdh — CHANGES_REQUESTED** [review]
Let me know after the changes has been made. Ill test more of the edge cases and ill try to simplify the whole code

**Comment by vmatsiiako** [comment]
Hey guys! Really excited about this PR! When are you thinking to merge this?

**Comment by sheensantoscapadngan** [comment]
Hey @Salman2301! I'd like to work on top of this and help you finally resolve this PR. I'll be pushing a few changes in a bit.

File: frontend/src/views/Settings/ProjectSettingsPage/components/EnvironmentSection/UpdateEnvironmentModal.tsx [hunk 11]
```diff
@@ -16,7 +16,7 @@ type Props = {
 
 const schema = yup.object({
   name: yup.string().label("Environment Name").required(),
-  slug: yup.string().label("Environment Slug").required()
+  slug: yup.string().label("Environment Slug").matches(/^[^./]*$/g, { message: "Invalid [.] or [/] not allowed"}).required()
```
Comment (sheensantoscapadngan) [inline frontend/src/views/Settings/ProjectSettingsPage/components/EnvironmentSection/UpdateEnvironmentModal.tsx:19]: use slugify

File: frontend/src/components/v2/InfisicalSecretInput/InfisicalSecretInput.tsx [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+        case ReferenceType.ENVIRONMENT:
+          replaceReference = `${currentPath}.`;
+          offset -= 1;
+          break;
+        default:
+      }
+      replaceReference = replaceReference.replace(/[//]/g, "");
+      newValue = `${start}$\{${replaceReference}}${end}`;
+      setValue(newValue);
+      // TODO: there should be a better way to do
```
Comment (sheensantoscapadngan) [inline frontend/src/components/v2/InfisicalSecretInput/InfisicalSecretInput.tsx:248]: complete this

File: frontend/src/components/v2/SecretInput/SecretInput.tsx (hunk 3 above)
Comment (sheensantoscapadngan) [reply frontend/src/components/v2/SecretInput/SecretInput.tsx]: Hey @akhilmhdh!  this regex is shared among the helper function `isValidSecretReferenceValue` and the `InfisicalSecretInput` component. 

The `isValidSecretReferenceValue` is used inside the secrets mutation file (hook). Would you happen to have a recommendation for how to structure this instead? 

image
image

**Comment by sheensantoscapadngan** [comment]
https://github.com/Infisical/infisical/assets/65645666/a61ee35e-7d2b-4721-88f3-e4902f721525

File: frontend/src/components/v2/InfisicalSecretInput/InfisicalSecretInput.tsx [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+              const referenceEndIndex = referenceStartIndex + (match?.[0]?.length || 0);
+              const [start, end] = [
+                value.slice(0, referenceStartIndex),
+                value.slice(referenceEndIndex)
+              ];
+
+              const newValue = start + end;
+              setValue(newValue);
+              onChange?.({ target: { value: newValue } } as any);
+            }
```
Comment (infisical-sheen) [inline frontend/src/components/v2/InfisicalSecretInput/InfisicalSecretInput.tsx:365]: move this to line 346

**Review by maidul98 — APPROVED** [review]
Great work!
